File: problem1.go
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
//TC - O(N)
//SC - O(N)

//BFS
func rightSideView(root *TreeNode) []int {
    if root == nil{
        return []int{}
    }
    
    result := []int{}
    q := []*TreeNode{}
    q = append(q,root)
    
    for len(q) > 0{
        size := len(q)    
        for i:=0;i<size;i++{
            node := q[0]
            q = q[1:]
            if i == size-1{
                result = append(result,node.Val)
            }
            
            if node.Left != nil{
                q = append(q,node.Left)
            }
            
            if node.Right != nil{
                q = append(q,node.Right)
            }
        }
    }
    
    return result
}


//DFS
var result []int

func rightSideView(root *TreeNode) []int {
    if root == nil{
        return []int{}
    }
    
    result = []int{}
    
    helper(root,0)
    
    return result
}


func helper(root *TreeNode,level int){
    if root == nil{
        return
    }
    
    if level == len(result){
        result = append(result,root.Val)
    }
    
    
    helper(root.Right,level+1)
    helper(root.Left,level+1)
    
    
}

File: Problem_1.js
// n is total nodes of tree
// h is the tree's height

// DFS:
// Time Complexity :
// O(n)
// Space Complexity:
// O(h)

// BFS:
// Time Complexity :
// O(n)
// Space Complexity:
// O(n)

// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Binary Tree Right Side View (https://leetcode.com/problems/binary-tree-right-side-view/)
// Your code here along with comments explaining your approach

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

// Post order traversal
let result;
let dfs = (root, level) => {
    // Base case
    if (root === null) return;

    // Logic
    if (result.length === level) {
        result.push(root.val);
    }
    dfs(root.right, level + 1);
    dfs(root.left, level + 1);
}

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function (root) {
    if (root === null)
        return [];

    // DFS
    result = [];
    dfs(root, 0);

    // BFS
    //     let q = [];
    //     let result = [];
    //     q.push(root);

    //     while(q.length > 0){
    //         let size = q.length;
    //         for(let i=0; i<size; i++){
    //             let node = q.shift();
    //             if(i === size-1)
    //                 result.push(node.val);

    //             if(node.left !== null)  q.push(node.left);
    //             if(node.right !== null)  q.push(node.right);
    //         }
    //     }

    return result;
};